akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala

```diff
@@ -514,46 +617,43 @@ private[akka] class Shard(
   private def receiveTerminated(ref: ActorRef): Unit = {
     if (handOffStopper.contains(ref))
       context.stop(self)
-    else if (idByRef.contains(ref) && handOffStopper.isEmpty)
-      entityTerminated(ref)
   }
 
   @InternalStableApi
-  def entityTerminated(ref: ActorRef): Unit = {
+  def entityTerminated(ref: ActorRef, id: EntityId): Unit = {
     import settings.tuningParameters._
-    val id = idByRef(ref)
-    idByRef -= ref
-    refById -= id
+    val isPassivating = entities.isPassivating(id)
```

Comment:
could be pushed to the else

Reply:
It can't as it has to be done before we mark it as terminated

Reply:
Ah, that's why!